### PR TITLE
CRC32 with SSE4.2

### PR DESCRIPTION
**Pull request by gperciva**

### CODE REVIEW

**Comment by gperciva** [comment]
Addresses #43.  Cuts the speed of the performance tests from 99 usec to 14 usec.  A bit hard to tell, since they're so fast and has a fair amount of variation.

This is a draft; I'm highlighting some specific questions to be resolved first.

File: alg/crc32c.c [hunk 1]
```diff
@@ -74,6 +83,67 @@ init(void)
...
+
+		/* Test case: "a". */
+		memset(block, 0, 4);
+		block[0] = 'a';
+		if (crctest(block)) {
+			warn0("Disabling hardware CRC due to failed self-test");
+			break;
+		}
+
+		/* SHA works; use it. */
```
Comment (gperciva) [inline alg/crc32c.c]: whoops, I just noticed this incorrect comment.

File: alg/crc32c_sse42.c [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+/**
+ * CRC32C_Update_SSE42(state, buf, len):
+ * Feed ${len} bytes from the buffer ${buf} into the CRC32C whose state is
+ * ${state}.  This implentation uses x86 SSE4.2 instructions, and should only
+ * be used if CPUSUPPORT_X86_CRC32 is defined and cpusupport_x86_crc32 returns
+ * nonzero.
+ */
+void
+CRC32C_Update_SSE42(uint32_t * state, const uint8_t * buf, size_t len)
```
Comment (gperciva) [inline alg/crc32c_sse42.c]: Question about naming this function, and the source file in general.  Following the pattern of other CPU-specific source files, we'd have `crc32c_crc32` which IMO looks goofy.  I went with `crc32c_SSE42`, but I'll be entirely unsurprised if you tell me to switch to `crc32c_crc32`.

File: tests/crc32/main.c [hunk 3]
```diff
@@ -20,6 +24,65 @@ static struct testcase {
 	    {0x1b, 0xc4, 0xb4, 0x28}}
 };
 
+/* Largest buffer must be first. */
+static const size_t perfsizes[] = {16384, 8192, 4096, 2048, 1024, 512, 256,
```
Comment (gperciva) [inline tests/crc32/main.c]: Should we add some more tests?  Or repeat these a few times?  it's hard to measure things that take less than 100 us.

File: alg/crc32c_sse42.c [hunk 4]
```diff
@@ -15,9 +17,15 @@
 void
 CRC32C_Update_SSE42(uint32_t * state, const uint8_t * buf, size_t len)
 {
-	const size_t block_len = len - len % 8;
+	const size_t pre = ((uintptr_t)buf & 7) < len ? (uintptr_t)buf & 7 : len;
+	const size_t remaining = ((len - pre) < len) ? len - pre : 0;
```
Comment (gperciva) [inline alg/crc32c_sse42.c]: This is rather ugly.  I'm quite curious to hear the proper way of making sure that we're loading values with proper alignment.

**Comment by gperciva** [comment]
Updated PR.

Hashing 100 mb in the different sizes takes ~1.7 seconds with SSE42, and ~2.0 seconds without.  I was expecting to see more of a difference, so I'm highly suspicious that something's wrong.

File: alg/crc32c_sse42.c [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+	size_t i = 0;
+
+	/* Process bytes before the alignment. */
+	for (; i < pre; i++)
+		*state = _mm_crc32_u8(*state, buf[i]);
+
+	/* Process blocks of 8. */
+	for (; i < block_len; i += 8) {
+		*state = (uint32_t)_mm_crc32_u64(*state,
+		    *(const unsigned long *)&buf[i]);
```
Comment (gperciva) [inline alg/crc32c_sse42.c]: This looks rather clumsy.  Is there a better way of getting an `(unsigned long long)` from `uint8_t` ?

File: alg/crc32c.c [hunk 6]
```diff
@@ -1,8 +1,17 @@
 #include <assert.h>
 #include <stdint.h>
 
+#include <warnp.h>
+
+#include "crc32c_sse42.h"
+#include "string.h"
```
Comment (cperciva) [inline alg/crc32c.c]: What's `"string.h"`?

File: tests/crc32/main.c [hunk 7]
```diff
@@ -20,6 +24,71 @@ static struct testcase {
...
+	size_t num_hashes;
+
+	/* Allocate buffer to hold largest message. */
+	if ((largebuf = malloc(perfsizes[0])) == NULL) {
+		warnp("malloc");
+		goto err0;
+	}
+	memset(largebuf, 0, perfsizes[0]);
+
+	/* Get beginning time. */
```
Comment (cperciva) [inline tests/crc32/main.c]: Get the start and end time for each size.  We want to say how fast each size of CRC32 takes.

File: tests/crc32/main.c [hunk 8]
```diff
@@ -20,6 +24,71 @@ static struct testcase {
...
+		goto err1;
+	}
+
+	/* Run operations. */
+	for (i = 0; i < num_perf; i++) {
+		num_hashes = bytes_to_hash / perfsizes[i];
+		CRC32C_Init(&ctx);
+
+		/* Hash all the bytes. */
+		for (j = 0; j < num_hashes; j++)
```
Comment (cperciva) [inline tests/crc32/main.c]: Loop `num_hashes` times and call `_Init` + `_Update` + `_Final` each time.  Most checksums we calculate will be of a single buffer; the initialization and finalization times matter.

File: tests/crc32/main.c [hunk 9]
```diff
@@ -52,8 +123,13 @@ main(int argc, char * argv[])
 		}
 	}
 
-	if (failures)
-		return (1);
-	else
-		return (0);
+	if (failures || performance_test())
```
Comment (cperciva) [inline tests/crc32/main.c]: I would either put the performance testing into a different utility or invoke it with a flag (like `test_sha256` does).  No reason why `make test` should run a performance test here -- it's not like mpool where we can verify that the performance is as expected.

**Comment by gperciva** [comment]
Updated PR, including a few tweaks to history.

I'm still concerned about the relatively insignificant speed change.  Doing ~500 MB is 0.6s with CRC32 instructions, and 0.7s without.  Granted, once the framework is in place, it would be easy to tweak the code in `alg/crc32c_sse42.c`.

With CRC32 CPU:
```
$ ./tests/crc32/test_crc32 -t
Hashing 5.37e+08 bytes...
... in 8068 blocks of size 66536:	0.63 s
... in 16384 blocks of size 32768:	0.58 s
... in 32768 blocks of size 16384:	0.59 s
... in 65536 blocks of size 8192:	0.58 s
... in 131072 blocks of size 4096:	0.58 s
... in 262144 blocks of size 2048:	0.59 s
... in 524288 blocks of size 1024:	0.59 s
... in 1048576 blocks of size 512:	0.58 s
... in 2097152 blocks of size 256:	0.59 s
... in 4194304 blocks of size 128:	0.58 s
... in 8388608 blocks of size 64:	0.58 s
... in 16777216 blocks of size 32:	0.58 s
... in 33554432 blocks of size 16:	0.71 s
```

without:
```
$ ./tests/crc32/test_crc32 -t
Hashing 5.37e+08 bytes...
... in 8068 blocks of size 66536:	0.79 s
... in 16384 blocks of size 32768:	0.70 s
... in 32768 blocks of size 16384:	0.70 s
... in 65536 blocks of size 8192:	0.70 s
... in 131072 blocks of size 4096:	0.71 s
... in 262144 blocks of size 2048:	0.71 s
... in 524288 blocks of size 1024:	0.73 s
... in 1048576 blocks of size 512:	0.73 s
... in 2097152 blocks of size 256:	0.71 s
... in 4194304 blocks of size 128:	0.72 s
... in 8388608 blocks of size 64:	0.71 s
... in 16777216 blocks of size 32:	0.74 s
... in 33554432 blocks of size 16:	0.75 s
```

**Comment by cperciva** [comment]
On 04/29/18 21:51, Graham Percival wrote:
> With CRC32 CPU:

Maybe a dumb question, but how are you doing this testing?  I mean, the
results are only meaningful if you're comparing the same CPU with and
without the CRC32 instructions.

> 8068 blocks of size 66536

You know that 66536 isn't a power of 2, right?

-- 
Colin Percival
Security Officer Emeritus, FreeBSD | The power to serve
Founder, Tarsnap | www.tarsnap.com | Online backups for the truly paranoid

**Comment by gperciva** [comment]
Whoops.  fixed 65536.

To test it without CRC32, I change `cpusupport/cpusupport_x86_crc32.c` to include:
```
        return 0 * ((ecx & CPUID_SSE42_BIT) ? 1 : 0);
```
since that's an easy way to disable it without having any "unused macros" or other stuff that triggers my paranoid-warnings compilation.

Granted, so far I haven't done any serious testing -- I mean, I haven't played games with CPU frequency scaling, or even simply canceling music playback.  That said, the results are very consistent for blocksizes other than 65536 and 16.

File: alg/crc32c_sse42.c [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+ * Feed ${len} bytes from the buffer ${buf} into the CRC32C whose state is
+ * ${state}.  This implentation uses x86 SSE4.2 instructions, and should only
+ * be used if CPUSUPPORT_X86_CRC32 is defined and cpusupport_x86_crc32 returns
+ * nonzero.
+ */
+void
+CRC32C_Update_SSE42(uint32_t * state, const uint8_t * buf, size_t len)
+{
+	const size_t pre = ((uintptr_t)buf & 7) < len ? (uintptr_t)buf & 7 : len;
+	const size_t remaining = ((len - pre) < len) ? len - pre : 0;
```
[the remark this answers is not in the document]
Comment (gperciva) [reply alg/crc32c_sse42.c]: "are there any bytes left after we've processed the 64-bit chunks?"

Simply doing `len - pre` could sometimes give me a negative number, which is a large `size_t` number.

[the remark this answers is not in the document]
Comment (gperciva) [reply alg/crc32c_sse42.c]: wait, no.  "are there any bytes left after we've processed the initial bytes"

... hmm, wait.  It doesn't make sense that we can end up with a negative value there.  Maybe my `pre` calculation is wrong?

[the remark this answers is not in the document]
Comment (gperciva) [reply alg/crc32c_sse42.c]: wow, replacing that with simply `len - pre` makes the CRC32 code much much faster.  And still passes the self-tests.

[the remark this answers is not in the document]
Comment (cperciva) [reply alg/crc32c_sse42.c]: Yeah.  In the case of an aligned buffer (aka. `pre = 0`) this was making it skip the 8-bytes-at-a-time loop completely...

**Comment by gperciva** [comment]
Rebased PR.  I took out the 64k and 32k versions since they weren't so interesting after an adequate warm-up.  Updated times.

With CRC32:
```
$ ./tests/crc32/test_crc32 -t
Hashing 5.37e+08 bytes...
... in 32768 blocks of size 16384:	0.10 s
... in 65536 blocks of size 8192:	0.08 s
... in 131072 blocks of size 4096:	0.07 s
... in 262144 blocks of size 2048:	0.07 s
... in 524288 blocks of size 1024:	0.07 s
... in 1048576 blocks of size 512:	0.06 s
... in 2097152 blocks of size 256:	0.06 s
... in 4194304 blocks of size 128:	0.07 s
... in 8388608 blocks of size 64:	0.12 s
... in 16777216 blocks of size 32:	0.27 s
... in 33554432 blocks of size 16:	0.48 s
```

without:
```
$ ./tests/crc32/test_crc32 -t
Hashing 5.37e+08 bytes...
... in 32768 blocks of size 16384:	0.72 s
... in 65536 blocks of size 8192:	0.70 s
... in 131072 blocks of size 4096:	0.70 s
... in 262144 blocks of size 2048:	0.70 s
... in 524288 blocks of size 1024:	0.70 s
... in 1048576 blocks of size 512:	0.71 s
... in 2097152 blocks of size 256:	0.70 s
... in 4194304 blocks of size 128:	0.71 s
... in 8388608 blocks of size 64:	0.70 s
... in 16777216 blocks of size 32:	0.72 s
... in 33554432 blocks of size 16:	0.75 s
```

**Pull request closed without merging**